src/app/credentials.rs

```diff
@@ -1,51 +1,86 @@
 use secret_service::{EncryptionType, Error, SecretService};
-use serde::{Deserialize, Serialize};
-use serde_json::{from_str, to_string};
+use serde::{de::Error as _, ser::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 use std::collections::HashMap;
-use std::str::from_utf8;
+use std::time::{Duration, SystemTime};
 
 static SPOT_ATTR: &str = "spot_credentials";
 
-#[derive(Deserialize, Serialize, Clone, Debug)]
-pub struct Credentials {
-    pub username: String,
-    pub password: String,
-    pub token: String,
-    pub country: String,
-}
-
 fn make_attributes() -> HashMap<&'static str, &'static str> {
     let mut attributes = HashMap::new();
     attributes.insert(SPOT_ATTR, "yes");
     attributes
 }
 
-pub fn try_retrieve_credentials() -> Result<Credentials, Error> {
-    let service = SecretService::new(EncryptionType::Dh)?;
-    let collection = service.get_default_collection()?;
-    let result = collection.search_items(make_attributes())?;
+mod serde_unix_timestamp {
+    use super::*;
 
-    let item = result.get(0).ok_or(Error::NoResult)?.get_secret()?;
-    let raw = from_utf8(&item).unwrap().to_string();
-    let parsed = from_str(&raw).map_err(|_| Error::Parse)?;
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<SystemTime>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        match <Option<u64>>::deserialize(deserializer)? {
+            Some(v) => Ok(Some(
+                SystemTime::UNIX_EPOCH
+                    .checked_add(Duration::from_secs(v))
+                    .ok_or_else(|| D::Error::custom("overflow deserializing SystemTime"))?,
+            )),
+            None => Ok(None),
+        }
+    }
 
-    Ok(parsed)
+    pub fn serialize<S>(system_time: &Option<SystemTime>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        system_time
+            .map(|v| match v.duration_since(SystemTime::UNIX_EPOCH) {
+                Ok(v) => Ok(v.as_secs()),
+                Err(_) => Err(S::Error::custom("SystemTime must be later than UNIX_EPOCH")),
+            })
+            .transpose()?
+            .serialize(serializer)
+    }
 }
 
-pub fn save_credentials(creds: Credentials) -> Result<(), Error> {
-    let service = SecretService::new(EncryptionType::Dh)?;
-    let collection = service.get_default_collection()?;
-    let encoded = to_string(&creds).unwrap();
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct Credentials {
+    pub username: String,
+    pub password: String,
+    pub token: String,
+    #[serde(with = "serde_unix_timestamp")]
```

Comment:
ah, I get it now!

maybe we should directly implement the traits rather than having this module? on a wrapper struct

Reply:
The traits `Serialize` and `Deserialize` can't be implemented directly since `SystemTime` already implements them. (The reason I wrote custom serialize/deserialize functions is because `SystemTime` normally also has a field for nano seconds that isn't relevant for us because the time returned by the Spotify API isn't that accurate).

The downside of using a wrapper struct is that it has to be used each time `Credentials` is constructed.

An alternative would be to move the serialize/deserialize functions out of the module and add `#[serde(serialize_with = "serialize_unix_timestamp", deserialize_with = "deserialize_unix_timestamp")]`. This way there isn't an extra module in the file and no wrapper type has to be used.

What do you prefer?

Reply:
Oh, missed that serde had an implementation for that already but you're right! Sure, we can't pretend to have it down to the nanoseconds, but for practical purposes I feel this would be fine to use the default serializer/deserializer?

Reply:
Yeah makes sense. Added a commit to use the default serializer/deserializer.